### PR TITLE
Logic for authentication added.

### PR DESCRIPTION
**Pull request by uzair-imtiaz**
### Description
This pull request marks the initial contribution to the project and introduces several key features to the authentication app. The changes include the addition of four templates and modifications to the **views.py** file, aimed at enhancing the user registration, login, and welcome functionalities of the Django application.

### Templates

1. **register.html**: This template serves as the user registration form, allowing users to input their registration details.

2. login.html: The login template presents a form for users to enter their login credentials.

3. **welcome.html**: This template displays a welcome message to authenticated users, providing a warm greeting upon successful login.

4. **base.html**: The base template establishes the layout for other templates by providing a common structure and elements that are extended by the remaining templates.

### Views
1. **RegisterView**: This class-based view manages the user registration process. It includes both the get() and post() methods to render the registration form and handle form submissions, respectively.

2. **LoginView**: Similar to the RegisterView, the LoginView is a class-based view responsible for user login functionality. It provides the get() and post() methods to handle rendering the login form and processing login submissions.

3. **welcome_view**: This function-based view is responsible for rendering the welcome message to authenticated users. It includes only the get() method, allowing users to receive a personalized welcome upon successful login.

### Testing
To test the newly added features and ensure their proper functioning, follow the steps below:

1. Clone the repository:
`git clone git@github.com:Uzair-Imtiaz/Code-Book.git`

2. Create a virtual environment:
`python3 -m venv myenv`

3. Activate the venv:
`source myenv/bin/activate`

4. Install Django:
`pip install Django`

2. Navigate to the project directory:
`cd Code-Book`

6. Apply database migrations:
`python3 manage.py migrate`

7. Start the development server:
`python3 manage.py runserver`

8. Open a web browser and visit the application at **http://localhost:8000/**.

9. Test the user registration functionality by accessing the registration form, entering valid registration details, and verifying that a new user is successfully created.

10. Test the user login functionality by accessing the login form, entering valid login credentials, and confirming that the user is logged in and redirected to the appropriate page.

11. Check the welcome message functionality by logging in as an authenticated user and ensuring that the welcome message is displayed on the relevant page.

### Snapshots
Screenshot 2023-07-05 at 11 44 24 AM

image

Screenshot 2023-07-05 at 11 51 21 AM

Screenshot 2023-07-05 at 11 51 55 AM

### CODE REVIEW

File: authentication/apps.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+from django.apps import AppConfig
+
+
+class AuthenticationConfig(AppConfig):
```
Comment (Ali-D-Akbar) [inline authentication/apps.py:15]: Docstring missing

File: authentication/templates/authentication/sign_in.html [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          <button type="button" class="btn btn-primary">
+              <a href="{% url 'authentication:register' %}" id="reg" style="color: #ffffff;text-decoration: none;">
+                  Create an Account
+              </a>
+          </button>
+      </form>
+    </div>
+  </div>
+
+{% endblock %}
```
Comment (Ali-D-Akbar) [inline authentication/templates/authentication/sign_in.html]: EoL missing

File: authentication/templates/authentication/welcome.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+{% block content %}
+
+  <div class="container">
+    <div class="welcome-message">
+      <h2 class="text-center mb-4">Welcome to the Site</h2>
+      <p class="text-center"> {{ user.first_name }} </p>
+    </div>
+  </div>
+
+{% endblock %}
```
Comment (Ali-D-Akbar) [inline authentication/templates/authentication/welcome.html]: EoL missing

File: authentication/views.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+from django.shortcuts import render, redirect, reverse
```
Comment (Ali-D-Akbar) [inline authentication/views.py:3]: Docstring missing

**Review by Ali-D-Akbar — COMMENTED** [review]
Some screenshots need to be updated after changes.

- [ ] To verify if we should use label and placeholder together

File: authentication/templates/authentication/sign_in.html (hunk 2 above)
Comment (uzair-imtiaz) [reply authentication/templates/authentication/sign_in.html]: Added.

File: authentication/templates/authentication/welcome.html (hunk 3 above)
Comment (uzair-imtiaz) [reply authentication/templates/authentication/welcome.html]: Added.

File: authentication/views.py (hunk 4 above)
Comment (uzair-imtiaz) [reply authentication/views.py:3]: Added.

File: authentication/static/authentication/css/register.css [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      align-items: center;
+      height: 100vh;
+    }
+
+    .form-container {
+      width: 400px;
+      padding: 40px;
+      background-color: #fffcfc;
+      border: 1px solid #000080;
+    }
```
Comment (Ali-D-Akbar) [inline authentication/static/authentication/css/register.css]: tabs can be removed.

File: authentication/templates/authentication/base.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{% load static %}
+
+<!DOCTYPE html>
+<html lang="en">
+<head>
```
Comment (Ali-D-Akbar) [inline authentication/templates/authentication/base.html:6]: one tab required.

File: authentication/templates/authentication/sign_in.html [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+{% block specific_css %}
+    <link rel="stylesheet" href="{% static 'authentication/css/signin.css' %}">
+{% endblock %}
+
+{% block content %}
+
+    {% if message is not None %}
+        <div class="container-top">
+            <div class="row">
+                <div class="col-3"></div>
```
Comment (Ali-D-Akbar) [inline authentication/templates/authentication/sign_in.html]: `<div class="col-3" />`

File: authentication/urls.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+from . import views
+
+app_name = 'authentication'
+
+urlpatterns = [
+    path('register/', views.RegisterView.as_view(), name='register'),
+    path('login/', views.LoginView.as_view(), name='login'),
+    path('welcome/', views.welcome, name='welcome')
+]
```
Comment (Ali-D-Akbar) [inline authentication/urls.py]: EoL missing

File: authentication/views.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+This file contains view classes and a function for handling user registration, login, and rendering the welcome page.
+
+Classes:
+- RegisterView(View): View class for user registration.
+- LoginView(View): View class for user login.
+
+Functions:
+- welcome(request): Renders the welcome page.
+"""
```
Comment (Ali-D-Akbar) [inline authentication/views.py]: """
Django views for authentication app
"""

**Review by Ali-D-Akbar — COMMENTED** [review]
Some readability can be improved. Use some settings in VScode or Pycharm to use 2 spaces tabs for frontend files.

File: authentication/tests.py [hunk 10]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (Ali-D-Akbar) [inline authentication/tests.py:3]: Docstring

File: projects/admin.py [hunk 11]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (Ali-D-Akbar) [inline projects/admin.py]: Docstrings to be added in projects app.

**Review by Ali-D-Akbar — APPROVED** [review]
Nits before final approval.

**Pull request merged**